keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,594 @@
+Once kubeadm moves out, utility functions will be exposed for public use,
+but perhaps without any guarantees until the kubeadm backend is truly
+announced for public consumption.
+
+### Implementation Details/Notes/Constraints
+
+#### Location
+
+The proposed new location is the kubernetes/kubeadm repository under
+a "kubeadm" folder with a sub-module.
+
+This makes it possible to use `go get k8s.io/kubeadm/kubeadm` to obtain
+the kubeadm source code. It also provides nice separation with the
+existing contents of the k/kubeadm repository such as "kinder", "operator"
+and various docs and tests.
+
+One side effect is that these contents will now be tagged and branched with
+kubeadm, but this is actually a positive change.
+
+#### Handling older kubeadm versions
+
+Currently kubeadm in the `release-1.17` branch (and later) of k/k
+does not depend on the rest of k/k. This was required for kubeadm
+to move out, however the older branches of k/k include versions of
+kubeadm that still depend on k/k.
+
+The proposed way forward is to move only the latest kubeadm from k/k master
+and build older version of kubeadm from k/k.
+
+This creates the following complications:
+- The post-submit job must handle tags and branches for kubeadm versions
+larger than X.
+- The kubernetes/release tooling must build kubeadm version X from the
+kubeadm repository while build older kubeadm from k/k.
+
+#### Releasing and tagging
+
+##### Challenges
+
+The current plan is for kubeadm to continue to follow the k8s version scheme
+and release schedule. This will continue to make the skew of kubeadm
+and k8s easy to understand for the users and continue to give motivation
+to the kubeadm developers to get work done on time.
+
+This creates some complications:
+- Automation tools have to build kubeadm as part of the release
+- The k/kubeadm tags and branches need to follow those of k/k
+- k/k tags are not SEMVER, which can cause troubles when importing
+the k/kubeadm repository
+
+Currently kubeadm is released as part of the following k8s artifacts:
+- Node tarballs (Linux / Windows)
+- deb/rpm packages
+
+The kubeadm binary is built using either bazel (amd64 only) or using Docker
+(cross-build, all supported architectures).
+
+The binary is built with the same version as the rest of the released binaries
+like the kubelet. The version tooling resides in k/k under the /hack directory
+and it must be copied to k/kubeadm for consistency. The tooling that perform
+the packaging and building are location in k/release and require manual interaction.
+
+##### Bazel vs go build
+
+Bazel was investigated as an option for building kubeadm from the new
+repository but the benefits are unclear. BUILD files present a maintenance
+challenge and also the bazel WORKSPACE and gazelle files need to be kept
+in sync with k8s. On the other hand building using “go build” can take
+around ~20 sec on a fast machine with fast internet.
```

Comment:
How is the RPM/DEB artifacts handled? Isn't `bazel` taking care of building those?

Reply:
no, currently those are built from specs in k/release.

Reply:
Currently Bazel is used only for CI DEB/RPM builds. The release ones are handled from k/release.

Reply:
and that is a problem because the DEB/RPM bazel build is assuming cmd/kubeadm is in k/k:
https://github.com/kubernetes/kubernetes/blob/master/build/debs/BUILD
(as much as i can read bazel)

in any case i don't think we should run bazel at all in the kubeadm repo, therefore this process may have to change.

Reply:
adding a topic about this in the Open Questions section.
my proposal is to stop building CI DEBs/RPMs from k/k as i haven't seen any users for these.